config/placeholder_replacement.go

```diff
@@ -144,18 +153,22 @@ func (v *placeholderReplacementVisitor) replacePlaceholders(s string) (string, e
 		}
 		placeholderKey := matches[placeholderRegexp.SubexpIndex("placeholder_key")]
 
+		var err error
+		var replacementResult string
 		// Now, match against every way we know of doing placeholder replacement
-		if packagePlaceholderRegexp.Match(placeholderKey) {
-			replaced, err := v.replacePackagePlaceholder(string(placeholderKey))
-			if err != nil {
-				replacementErrors = multierr.Append(replacementErrors, err)
-				return placeholder
-			}
-			return []byte(replaced)
+		switch {
+		case packagePlaceholderRegexp.Match(placeholderKey):
+			replacementResult, err = v.replacePackagePlaceholder(string(placeholderKey))
+		case environmentPlaceholderRegexp.Match(placeholderKey):
+			replacementResult, err = v.replaceEnvironmentPlaceholder(string(placeholderKey))
+		default:
+			err = errors.Errorf("invalid placeholder %q", string(placeholder))
```

Comment:
I think this is much cleaner and should be easy for people to read / modify.

I tried your idea of splitting into a separate file, but the two files didn't make sense without each other but I don't think it ended up better -- the functions were very codependent.

Reply:
what was the dependency between functions? I am not super fussed on this because this definitely looks better but I just feel like having a readable function rather than the match code directly still would be better. Can refactor later as this placeholder stuff grows though.